From f2986ebea8a75f9697867c895ab49b6b112f7b65 Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Mon, 15 Sep 2025 13:31:26 +0200
Subject: a11y: Consider modal QML Popup for a11y hierarchy

Introduce QAccessibleQuickPopupItem to allow access to the a11y properties
of the QML Popup. This uses the QQuickPopupItem as QQuickPopup does not
derive from QQuckItem and the popup item is the actual parent of the
popup's content.

Via QAccessibleQuickPopupItem it is possible to set the `modal` flag
of QAccessible::State and use this to decide if the application is blocked
by a modal dialog. In this case only the top-most popup is exposed to the
a11y layer.

As a fly-by this commit removes redudant curly braces in
qtquicktemplates2global.cpp.

Fixes: QTBUG-140200
Change-Id: Id4a89894bc0576410b8a184dd6f18bfb1c8b2b8f
---
 src/quick/accessible/qaccessiblequickitem.cpp |  8 ++++++
 src/quicktemplates/CMakeLists.txt             |  1 +
 .../accessible/qaccessiblequickpopupitem.cpp  | 26 +++++++++++++++++++
 .../accessible/qaccessiblequickpopupitem_p.h  | 22 ++++++++++++++++
 .../qtquicktemplates2global.cpp               |  8 ++++--
 5 files changed, 63 insertions(+), 2 deletions(-)
 create mode 100644 src/quicktemplates/accessible/qaccessiblequickpopupitem.cpp
 create mode 100644 src/quicktemplates/accessible/qaccessiblequickpopupitem_p.h

diff --git x/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp y/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp
index db2249ecd81a091d9f48cc5239b232099292777b..1bc956b1d50af51526b11e5e304b54f8c217c778 100644
--- x/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp
+++ y/qtdeclarative/src/quick/accessible/qaccessiblequickitem.cpp
@@ -420,6 +420,14 @@ QList<QQuickItem *> accessibleUnignoredChildren(QQuickItem *item, bool paintOrde
 {
     QList<QQuickItem *> items;
     unignoredChildren(item, &items, paintOrder);
+
+    auto it = std::find_if(items.rbegin(), items.rend(), [](QQuickItem *item) {
+        const QAccessibleInterface *iface = QAccessible::queryAccessibleInterface(item);
+        return iface && iface->state().modal && iface->role() == QAccessible::Dialog;
+    });
+    if (it != items.rend())
+        return { *it };
+
     return items;
 }
 
diff --git x/qtdeclarative/src/quicktemplates/CMakeLists.txt y/qtdeclarative/src/quicktemplates/CMakeLists.txt
index 64be0cbe7cf8e8dbfa5f14b8b57fd7d3592567b1..644ad1afd7472201a42b0344d730e719fb0fa286 100644
--- x/qtdeclarative/src/quicktemplates/CMakeLists.txt
+++ y/qtdeclarative/src/quicktemplates/CMakeLists.txt
@@ -142,6 +142,7 @@ qt_internal_extend_target(QuickTemplates2 CONDITION TARGET Qt::QmlModels
 qt_internal_extend_target(QuickTemplates2 CONDITION QT_FEATURE_accessibility
     SOURCES
         accessible/qaccessiblequickpage.cpp accessible/qaccessiblequickpage_p.h
+        accessible/qaccessiblequickpopupitem.cpp accessible/qaccessiblequickpopupitem_p.h
 )
 
 qt_internal_extend_target(QuickTemplates2 CONDITION QT_FEATURE_quicktemplates2_container
diff --git x/qtdeclarative/src/quicktemplates/accessible/qaccessiblequickpopupitem.cpp y/qtdeclarative/src/quicktemplates/accessible/qaccessiblequickpopupitem.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..818cfec605d45459bbbcd9fa5f32d3c6e38d7855
--- /dev/null
+++ y/qtdeclarative/src/quicktemplates/accessible/qaccessiblequickpopupitem.cpp
@@ -0,0 +1,26 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "qaccessiblequickpopupitem_p.h"
+#include "qquickpopup_p.h"
+#include "qquickpopupitem_p_p.h"
+
+QT_BEGIN_NAMESPACE
+
+QAccessibleQuickPopupItem::QAccessibleQuickPopupItem(QQuickPopupItem *popupItem)
+    : QAccessibleQuickItem(popupItem)
+{
+}
+
+QAccessible::State QAccessibleQuickPopupItem::state() const
+{
+    QAccessible::State state = QAccessibleQuickItem::state();
+
+    QQuickPopup *popup = static_cast<QQuickPopup *>(object()->parent());
+    if (popup)
+        state.modal = popup->isModal();
+
+    return state;
+}
+
+QT_END_NAMESPACE
diff --git x/qtdeclarative/src/quicktemplates/accessible/qaccessiblequickpopupitem_p.h y/qtdeclarative/src/quicktemplates/accessible/qaccessiblequickpopupitem_p.h
new file mode 100644
index 0000000000000000000000000000000000000000..68580cf58de40ad93b860276632cf8113d54e153
--- /dev/null
+++ y/qtdeclarative/src/quicktemplates/accessible/qaccessiblequickpopupitem_p.h
@@ -0,0 +1,22 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#ifndef QACCESSIBLEQUICKPOPUPITEM_H
+#define QACCESSIBLEQUICKPOPUPITEM_H
+
+#include <QtQuick/private/qaccessiblequickitem_p.h>
+
+QT_BEGIN_NAMESPACE
+
+class QQuickPopupItem;
+
+class QAccessibleQuickPopupItem : public QAccessibleQuickItem
+{
+public:
+    QAccessibleQuickPopupItem(QQuickPopupItem *popupItem);
+    QAccessible::State state() const override;
+};
+
+QT_END_NAMESPACE
+
+#endif // QACCESSIBLEQUICKPOPUPITEM_H
diff --git x/qtdeclarative/src/quicktemplates/qtquicktemplates2global.cpp y/qtdeclarative/src/quicktemplates/qtquicktemplates2global.cpp
index e65eba230ee77e4c22da00fe1af4cade09542445..1db14d5038bd0f0ed34c6165676118fc32b49f7b 100644
--- x/qtdeclarative/src/quicktemplates/qtquicktemplates2global.cpp
+++ y/qtdeclarative/src/quicktemplates/qtquicktemplates2global.cpp
@@ -7,7 +7,9 @@
 
 #if QT_CONFIG(accessibility)
 #include "qquickpage_p.h"
+#include "qquickpopupitem_p_p.h"
 #include "accessible/qaccessiblequickpage_p.h"
+#include "accessible/qaccessiblequickpopupitem_p.h"
 #endif
 
 QT_BEGIN_NAMESPACE
@@ -15,9 +17,11 @@ QT_BEGIN_NAMESPACE
 #if QT_CONFIG(accessibility)
 static QAccessibleInterface *qQuickAccessibleFactory(const QString &classname, QObject *object)
 {
-    if (classname == u"QQuickPage") {
+    if (classname == u"QQuickPage")
         return new QAccessibleQuickPage(qobject_cast<QQuickPage *>(object));
-    }
+    if (classname == u"QQuickPopupItem")
+        return new QAccessibleQuickPopupItem(qobject_cast<QQuickPopupItem *>(object));
+
     return nullptr;
 }
 #endif
